src/wallet/asyncrpcoperation_sendmany.cpp

```diff
@@ -368,36 +387,39 @@ uint256 AsyncRPCOperation_sendmany::main_impl() {
                 builder_.SendChangeTo(changeAddr.value(), ovks.first);
             }
         },
+        [&](const libzcash::UnifiedAddress& ua) {
+            allowChangeTypes(ua.GetKnownReceiverTypes());
+
+            auto zufvk = pwalletMain->GetUFVKForAddress(ua);
+            if (!zufvk.has_value()) {
+                throw JSONRPCError(
+                        RPC_WALLET_ERROR,
+                        "Could not determine full viewing key for unified address.");
+            }
+
+            auto changeAddr = zufvk.value().GetChangeAddress(allowedChangeTypes);
+            if (!changeAddr.has_value()) {
+                throw JSONRPCError(
+                        RPC_WALLET_ERROR,
+                        "Could not generate a change address from the inferred full viewing key.");
+            }
+            builder_.SendChangeTo(changeAddr.value(), ovks.first);
+        },
         [&](const libzcash::UnifiedFullViewingKey& fvk) {
+            allowChangeTypes(fvk.GetKnownReceiverTypes());
             auto zufvk = ZcashdUnifiedFullViewingKey::FromUnifiedFullViewingKey(Params(), fvk);
-            auto changeAddr = zufvk.GetChangeAddress();
+            auto changeAddr = zufvk.GetChangeAddress(allowedChangeTypes);
             if (!changeAddr.has_value()) {
                 throw JSONRPCError(
                         RPC_WALLET_ERROR,
-                        "Could not generate a change address from the specified full viewing key ");
+                        "Could not generate a change address from the specified full viewing key.");
             }
             builder_.SendChangeTo(changeAddr.value(), ovks.first);
         },
         [&](const AccountZTXOPattern& acct) {
-            for (ReceiverType rtype : acct.GetReceiverTypes()) {
-                switch (rtype) {
-                    case ReceiverType::P2PKH:
-                    case ReceiverType::P2SH:
-                        allowedChangeTypes.insert(OutputPool::Transparent);
-                        break;
-                    case ReceiverType::Sapling:
-                        allowedChangeTypes.insert(OutputPool::Sapling);
-                        break;
-                    case ReceiverType::Orchard:
-                        if (builder_.SupportsOrchard()) {
-                            allowedChangeTypes.insert(OutputPool::Orchard);
-                        }
-                        break;
-                }
-            }
-
+            allowChangeTypes(acct.GetReceiverTypes());
             auto changeAddr = pwalletMain->GenerateChangeAddressForAccount(
-                        acct.GetAccountId(),
+                        sendFromAccount_,
```

Comment:
I presume that `sendFromAccount_` is set to `acct.GetAccountId()`.

Reply:
That's correct.

Reply:
Hm, although it's probably better to use acct.GetAccountId() directly anyway.